Support transitions on widget parent changes

flackr opened the pull request:
This adds support for transitions from a previous position when a widget parent changes. I created a table with transitions on cards (while not dragged) where you can compare the behavior:

Compare that to the PR test here to see the difference.
Before: https://virtualtabletop.io/holder-transitions
With PR: https://test.virtualtabletop.io/PR-1423/pr-test

This is based on the PR in #1421 since it requires some of the newly added geometry helper methods in order to compute the relative transforms. I believe this fixes #508.

ArnoldSmith86 commented:
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1423/pr-test (or any other room on that server)

After merging, a backup will be available at `https://beta.virtualtabletop.io/editor/PR1423-pr-test`.

96LawDawg commented:
This looks amazing.  A question and a comment.

Question: how can we tweak the effect?  I see you put a "classes":"transitions" into the cardDefaults.  What css magic do I need to do to change the type of transition or change the length, for example?

Comment:  When dealing from the deck, sometimes the cards just appear without the transition.  Not sure if that is because more than one widget is moving at the same time or maybe for some other reason.  The missing transition animation is worse is Chrome than Edge.

flackr commented:
This works in 3d too, though the transition from a 2d transform to a 3d transform within perspective is a bit wonky:
https://test.virtualtabletop.io/PR-1423/perspective

flackr commented:
> Question: how can we tweak the effect? I see you put a "classes":"transitions" into the cardDefaults. What css magic do I need to do to change the type of transition or change the length, for example?

Check out the `css` widget, it's just a transition on the transform property defined there, specifically:
```
  "css": {
    "_, .transition": {
      "transition": "transform 150ms"
    },
    "_, .transition.dragging": {
      "transition": "none"
    }
  }
```
You can fully customize the length, easing curve, delays (which would be useful when dealing multiple cards to stagger the later cards) using all the standard css transition properties. Note that I specifically don't have a transition while dragging because that would make the card lag behind your drag.

> Comment: When dealing from the deck, sometimes the cards just appear without the transition. Not sure if that is because more than one widget is moving at the same time or maybe for some other reason. The missing transition animation is worse is Chrome than Edge.

Interesting, I hadn't seen this myself yet but I'll have to experiment more.

ArnoldSmith86 commented:
I'll have a closer look when #1421 is merged and this one updated. Too lazy to compare them.

96LawDawg commented:
Did a little testing after the commits since my last test.  Still some skipping of the visible effect when using the Deal button in the demo room.

One other thing, you might want to consider is the layer.  Not sure if it is possible to easily fix this.  But when I use the Move Up/Down buttons in your demo room, the cards slide behind the buttons and basic widgets in the room.  I slowed the animation way down to watch.  This is because the cards are in a holder which is on layer -3 and the buttons are -1 and the basic widget -2.  If you were manually dragging the cards, they would pass over the buttons instead of behind them.

I could probably correct for that in the JSON by putting a temp layer on the cards while they are moving.  But do you think this is solvable in what you are coding?

RaphaelAlvez commented:
Is there any reason why the deal doesn't looks as good as the move up and move down?

Also I would prefer having just a little bit slower

96LawDawg commented:
> Also I would prefer having just a little bit slower

The timing will be set by each game designer in the css.  That isn't baked into the code.

ArnoldSmith86 commented on client/js/serverstate.js:
```diff
@@ -100,10 +100,6 @@ export function batchEnd() {
 
 function receiveDelta(delta) {
   // the order of widget changes is not necessarily correct and in order to avoid cyclic children, this first moves affected widgets to the top level
-  for(const widgetID in delta.s)
-    if(delta.s[widgetID] && delta.s[widgetID].parent !== undefined && widgets.has(widgetID))
-      $('#topSurface').appendChild(widgets.get(widgetID).domElement);
-
```
Removing this breaks for example FreeCell ( https://test.virtualtabletop.io/PR-1423/freecell ). The 8 columns of widgets are shuffled and each card is a child of the one under it.

So when resetting the game, it basically every time causes a child-parent-relationship to swap. When applying the delta without first moving the `domElement`s to an intermediate parent, it creates circular parent relationships.

flackr replied:
Ah I see, to restate to make sure I'm looking at the same problem, the root issue is that given:
```html
<div id="a">
  <div id="b"></div>
</div>
```
```js
b.appendChild(a);
```
Will throw `DOMException: Failed to execute 'appendChild' on 'Node': The new child element contains the parent.`
Demo: https://jsbin.com/jejuvud/edit?html,css,js,output

I guess the most straightforward fix would be to give the widget the relative transform in `#topSurface` that it should have to keep it's original from position.

flackr replied:
Restored this in the latest patch, but adding in preserving the transform relative to the root so that the from position is calculated correctly in `widget.js` when adding to the new parent.

flackr commented:
> Is there any reason why the deal doesn't looks as good as the move up and move down?

As far as I can tell it's just that the card moves too fast to be seen. If I make the transition duration longer it works perfectly. Keep in mind this PR is allowing developers to use standard CSS transitions, it does not define how they use those transitions.

There have been proposals in the CSSWG for distance-based transition timing which would solve this in the browser css space, but we could explore implementing some version of this in a followup.

robartsd commented:
> As far as I can tell it's just that the card moves too fast to be seen.

I believe this can happen with PCIO's transitions as well.

96LawDawg reviewed (approved):
Approved through test.

I created the original timing based transition routine for my poker room.  So I tried this new css based version out.  It works great and is so much simpler.

96LawDawg commented:
I think I would also like to recommend that this effect be added by default to new decks with a transition time of maybe 700ms.  That's what I did in Hearts in https://test.virtualtabletop.io/PR-1423/lawdawg.  It helps visually to see what is going on with the cards.  Be sure to try sorting and manually dragging and dropping into your hand as well.

96LawDawg reviewed (changes requested):
I have to revoke my approval.  FreeCell still doesn't work.

ArnoldSmith86 commented:
And we all know that Jan gets angry when FreeCell doesn't work!

flackr commented:
@ArnoldSmith86 @96LawDawg Freecell works now. The issue was that we have to force the transform to be recomputed due to it being locally changed in `serverstate.js` but not updated by applyCSS due to not being part of the delta.

96LawDawg reviewed (approved):
Here.  Have another approval.

96LawDawg commented:
@flackr Here is a revised CSS tutorial that includes your transition demo and perspective demo.  The perspective demo didn't work because that branch wasn't in here, so I brought this branch up to date.  Now all of your recent changes have tutorials supporting them.  Thanks again for making VTT even better looking.


[CSS.zip](https://github.com/ArnoldSmith86/virtualtabletop/files/10357064/CSS.zip)

96LawDawg commented:
Well, that is just strange.  The last tutorial (Transform) works on production but does not work in this PR.  Could there be any conflict between them?  I don't understand what is going on.

flackr commented:
> Well, that is just strange. The last tutorial (Transform) works on production but does not work in this PR. Could there be any conflict between them? I don't understand what is going on.
> 
> See https://test.virtualtabletop.io/PR-1423/lawdawg

Looks about right now. I added the cardDefaults class and supporting CSS for the transitions and fixed the 3d transform on the table.

96LawDawg commented:
Well, I thought I copied the room exactly and just changed the text at the top.  Must have messed something up.  Save that (without the basic css timing widget--that makes it strange) as the last variant and put that in your PR.  Thanks for cleaning up the mess.

flackr commented:
> Well, I thought I copied the room exactly and just changed the text at the top. Must have messed something up. Save that (without the basic css timing widget--that makes it strange) as the last variant and put that in your PR. Thanks for cleaning up the mess.

No I think you got it right. It seems that the 3d css with the `!important` doesn't get applied until you trigger a change. I'll look into it but it seems like a bug (unrelated to this PR).

flackr commented:
@96LawDawg I fixed the tutorial loading in the latest update. I guess when applying the css inline it's overridden by the transform applied to set the widget position but if you set it with a stylesheet css (e.g. an object) then it is applied properly.

96LawDawg commented:
I think I figured out why I can't see your branches (because I'm not looking at your fork). But I'm such a n00b I can't figure out how to do that and I don't want to mess up what I have.  So here is one more tutorial change for you to please add to this PR.  I have added a comment to variant 1 of the Timer tutorial not to use that technique.  This was how I originally created card transition animations.  The capability demo is still useful as a timer tutorial, but the note advises people to use the technique you created here.  Thanks.

[Timer.zip](https://github.com/ArnoldSmith86/virtualtabletop/files/10361354/Timer.zip)

robartsd commented:
> I guess when applying the css inline it's overridden by the transform applied to set the widget position but if you set it with a stylesheet css (e.g. an object) then it is applied properly.

Yes, I believe setting element.style.transform will overwrite any inline transform styles. I think in production this only happens when a widget is moved (the initial transform is simply loaded when the inline styles are built). I think this PR may be setting it after loading css when initializing the widget

flackr commented:
> > I guess when applying the css inline it's overridden by the transform applied to set the widget position but if you set it with a stylesheet css (e.g. an object) then it is applied properly.
> 
> Yes, I believe setting element.style.transform will overwrite any inline transform styles. I think in production this only happens when a widget is moved (the initial transform is simply loaded when the inline styles are built). I think this PR may be setting it after loading css when initializing the widget

FWIW I don't think the issue was part of this PR, just part of @96LawDawg's tutorial based on my 3d css demo which was setting it using the inline css property. I don't think it's any different on tip of tree vs in this PR.

robartsd commented:
> FWIW I don't think the issue was part of this PR, just part of @96LawDawg's tutorial based on my 3d css demo which was setting it using the inline css property. I don't think it's any different on tip of tree vs in this PR as we already set style.transform here https://github.com/ArnoldSmith86/virtualtabletop/blob/main/client/js/widgets/widget.js#L146

That only gets called if a delta contains transform properties (parent, x, y) and does not contain any other css properties. If the delta contains any css properties that are not transform properties, the full css() method gets called instead (note the return a few lines above), so it doesn't get called until the widget gets moved.

```
  applyCSS(delta) {
    for(const property of this.classesProperties()) {
      if(delta[property] !== undefined) {
        this.domElement.className = this.classes();
        break;
      }
    }

    for(const property of this.cssProperties()) {
      if(delta[property] !== undefined) {
        this.domElement.style.cssText = mapAssetURLs(this.css());
        return;
      }
    }

    for(const property of this.cssTransformProperties()) {
      if(delta[property] !== undefined) {
        this.domElement.style.transform = this.cssTransform();
        return;
      }
    }
  }
```

flackr commented:
@robartsd sorry I explained myself poorly. I realize that it is technically a difference with this PR on initially loading a widget that it now sets the transform. My point was that given that the transform breaks if we ever update a transform related property later, it seems more consistent to have the initial state be consistent with what would happen if there's a later position update.

flackr commented:
That said if we want to preserve the current behavior I think it will work if we read the inline transform style before overriding it and then reapply that. Let me know what you think.

robartsd commented:
I personally don't care about changing this behavior.

Currently widget css is calculated and game engine transform is added to the end of the inline style string. If the widget css uses !important, it can override the engine transform with an inline style. If the game designer supported movement by using the widget transform properties in their css, then the properties would be in the array returned by this.cssProperties() and element.style.transform would never be set by the game engine. With this PR, this still works with widget stylesheets, but now fails with inline widget styles (I imagine that overwriting transform with !important in a stylesheet may cause this PR's support for transitions to fail).

I personally don't care if this breaks existing games that use !important to override the game engine transform.

96LawDawg commented:
> I personally don't care if this PR breaks existing games that use !important to override the game engine transform.

I do.  This breaks Crossed (the first dice no longer rolls).  It breaks variant 2 of the Dice Assets (so it probably breaks other things).  That's enough to require this to be reworked.  Our motto is not unlike that of doctors: above all else, do no harm -- we don't break existing games if we know about it.

96LawDawg reviewed (changes requested):
Need to fix issues that cause games to break.

robartsd commented:
A file updater that converts inline transform css to widget stylesheet transform css should fix this for most cases. The remaining issue to this approach would be inline css set by a routine.

flackr commented:
I didn't realize games were relying on this. It seems fragile, but I can see three possible approaches,
1. As @robartsd suggested, we could use a fileupdater to make the same modification that I did to the perspective tutorial.
2. Try to remember the previous `style.transform`. This is rather ugly as we'd need to detect whether it was changed by applyCSS and track it across files as it's updated in two locations.
3. Call applyCSS with a delta of `this.state` filtered to cssProperties(), and adding cssTransformProperties() to ensure that we reset the transform if nothing else does.

I would prefer not to do 2, but 1 or 3 both seem okay to me. 3 seems slightly more expensive but would preserve the current behavior, so I'd lean towards 1 - a file update for existing games.

flackr commented:
But ... parsing css for the file updater is non-trivial to get right, so I figured out a relatively simple way to track what the inline transform should be after the transition - just saving it after related changes in applyCSS.

flackr commented:
@96LawDawg I switched the tutorial back to what it was originally and verified this is all working.

96LawDawg reviewed (approved):
Everything works again.  Approved through test.

robartsd commented:
> But ... parsing css for the file updater is non-trivial to get right, so I figured out a relatively simple way to track what the inline transform should be after the transition (i.e. option 2) - just saving it after related changes in applyCSS.

This looks like a great solution to me.

robartsd reviewed (approved):
Code looks good.
